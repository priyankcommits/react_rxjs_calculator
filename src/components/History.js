// Component to show result history
// Listens to output box for results and Publishes to Inputbox for history selection
import Button from 'react-bootstrap/Button';
import ListGroup from 'react-bootstrap/ListGroup';
import Modal from 'react-bootstrap/Modal';
import React, { useState, useReducer, useEffect } from 'react';
import { skip } from 'rxjs/operators';

import { resultEvent$, resultSelectionEvent$ } from '../events';

const initialState = {
  entries: [],
}

const reducer = (state, action) => {
  switch(action.type) {
    case 'add':
      return {entries: [...state.entries, action.value]};
    case 'delete':
      let tempStateEntries = [...state.entries];
      tempStateEntries.splice(action.index, 1);
      return {entries: [...tempStateEntries]};
    default:
      return {...state};
  }
}

function History() {
  const [show, setShow] = useState(false);
  const handleClose = () => setShow(false);
  const handleShow = () => setShow(true);

  const [results, setResults] = useReducer(reducer, initialState);

  const selectEntry = (value) => {
    resultSelectionEvent$.next(value);
    handleClose();
  }

  const deleteEntry = (entryIndex) => {
    setResults({index: entryIndex, type: 'delete'});
  }

  useEffect(() => {
    resultEvent$.pipe(skip(1)).subscribe(data => {
      setResults({value: data, type: 'add'});
    });
    return () => {
      resultEvent$.unsubscribe();
    }
  }, []);

  return (
    <>
      <Button variant='dark' onClick={handleShow} size='sm' className='mt-1'>
        h
      </Button>

      <Modal show={show} onHide={handleClose}>
        <Modal.Header closeButton>
          <Modal.Title>
            Results History
          </Modal.Title>
        </Modal.Header>
        <Modal.Body>
          <ListGroup>
            {results.entries.map((result, index) => (
              <ListGroup.Item key={index}>
                {result.input} : {result.value}
                <Button variant='primary' size='sm' className='ml-2' onClick={() => selectEntry(result.value)}>Select</Button>
                <Button variant='danger' size='sm' className='ml-2' onClick={() => deleteEntry(index)}>Delete</Button>
              </ListGroup.Item>
            ))}
          </ListGroup>
        </Modal.Body>
      </Modal>
    </>
  )
}

export default History;