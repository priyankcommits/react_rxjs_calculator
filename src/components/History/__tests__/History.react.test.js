// Snapshot testing
import React from 'react';
import renderer from 'react-test-renderer';

import { History } from '../History';

test('History Layout Initial Rendering', () => {
  const component = renderer.create(
    <History />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

});
