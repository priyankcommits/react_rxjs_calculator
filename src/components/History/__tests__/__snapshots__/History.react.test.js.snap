// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`History Layout Initial Rendering 1`] = `
<div
  className="sc-AxhCb kgMfC mt-1 btn btn-dark btn-sm"
  disabled={false}
  onClick={[Function]}
>
  <span>
    H
  </span>
</div>
`;
